### DISCUSSION TITLE
Send an email without sendmail.

### DISCUSSION THREAD

**Discussion by WhiteTiger-IT** [fail2ban/fail2ban#3681] [General]
Is there a way to configure Fail2Ban to send the email without having to install a mail server on this PC?
That is, using an external SMTP server, for example that of my provider which I already use to send my personal email and other alerts.

**Comment by sebres** [comment]
Yes, see for instance [action.d/smtp.py](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/smtp.py)...
(just only newest version support SSL, so better update the action from link above).

One could also configure `sendmail`, `mail` and co to send mails through external SMTP relay.

**Reply by WhiteTiger-IT** [reply]
Thanks for the reply.
I wanted to avoid installing sendmail or other software.

Should the script be inserted into actiobn.d to replace the existing one?

**Comment by WhiteTiger-IT** [comment]
I renamed the smtp.py file to smtp-ok.py
Then I created a new smtp.py file by placing the contents of the file at your link.

In jail.local mtp = smtp, right?
Do I have to modify the data **host, ssl, user, password, sendername, sender, dest** in smtp.py or can I insert them under [sshd]?

**Reply by sebres** [reply]
> In jail.local mtp = smtp, right?

No. It is `smtp.py` (`.py` is required for pythonic actions).

> Do I have to modify the data host, ssl, user, password, sendername, sender, dest in smtp.py or can I insert them under [sshd]?

Yes you can specify them within jail section or default:
```ini
action = %(action_)s
         smtp.py[dest="account-name@example.com", host="smtp.example.com", ssl=1, user="account-name", password="account-pwd"]
```
IIRC, you can also add `, matches="matches"` to include matched lines to the mail.

`%(action_)s` would substitute default action (mostly `banaction` or `banaction_allports` only, depending on jail).

**Comment by WhiteTiger-IT** [comment]
I'm doing some tests, but no emails are arriving.
In action.d there is a smtp.py file with the content copied from your link.
In [sshd] should I use **action = %(action_)s** or **action = %(action_mwl)s** ?
With the second option systemctl restart gives an error.

This is my jail.local

```
## Modified: 27-02-2024

[INCLUDES]
before = paths-debian.conf

[DEFAULT]
ignoreip = 127.0.0.1/8 192.168.101.33 ::1
ignorecommand =

bantime  = 10m
findtime  = 10m
maxretry = 3
maxmatches = %(maxretry)s

backend = systemd

usedns = warn
logencoding = auto
enabled = false
mode = normal

filter = %(__name__)s[mode=%(mode)s]

destemail = root@localhost
sender = root@<fq-hostname>
mta = smtp.py
protocol = tcp

chain = <known/chain>

port = 0:65535

fail2ban_agent = Fail2Ban/%(fail2ban_version)s
banaction = iptables-multiport
banaction_allports = iptables-allports
action_ = %(banaction)s[port="%(port)s", protocol="%(protocol)s", chain="%(chain)s"]
action_mw = %(action_)s
            %(mta)s-whois[sender="%(sender)s", dest="%(destemail)s", protocol="%(protocol)s", chain="%(chain)s"]
action_mwl = %(action_)s
             %(mta)s-whois-lines[sender="%(sender)s", dest="%(destemail)s", logpath="%(logpath)s", chain="%(chain)s"]
action = %(action_mwl)s

[sshd]
#mode   = normal
enabled = true
maxretry  = 3
findtime  = 1d
bantime   = 1m
port    = ssh                    ; Port 22 (standard)
logpath = %(sshd_log)s
#backend = %(sshd_backend)s        ; For OSes other than Debian
backend = systemd                          ; For Debian
action = %(action_)s
         smtp.py[dest="my-email@mydomain.tld", host="smtp.my-isp.tld:465", ssl=1, user="mailsrv@mydomain.tld", password="MY-PSW",sendername="Fail2Ban", sender="pc-sys@otherdomain.tld"]
```

**Reply by sebres** [reply]
`mta = smtp.py` is wrong (also all actions definitions for `action*` in default section), because there is no actions like `smtp.py-whois` or `smtp.py-whois-lines`.
Only last two lines with action (the definition in sshd section) looks good. 

> In [sshd] should I use action = `%(action_)s` or `action = %(action_mwl)s` ?
> With the second option systemctl restart gives an error.

Yes, because if you have this:
```ini
[DEFAULT]
action_mwl = %(action_)s
             %(mta)s-whois-lines[sender="%(sender)s", dest="%(destemail)s", logpath="%(logpath)s", chain="%(chain)s"]
action = %(action_mwl)s
[other-jail]
enabled = true
```
the resulting jail-action will look like:
```
[other-jail]
action = iptables-multiport[...]
         smtp.py-whois-lines[...]
```
and as already said, there is no such action `smtp.py-whois-lines.conf` at all.

There is only `smtp.py` action.
One would need to implement `smtp-whois-lines.py` action to use it (but with direct name, since this default jail-config only useable for mta-related command actions (not for pythonic actions).

> I'm doing some tests, but no emails are arriving.

1. you must always restart fail2ban or jail (not reload) to update action-related parameters.
2. restored tickets (bans of still active intruders after restart) will not cause invocation of mailing action (since they are `norestored = true`), so it'd be invoked only for new fresh bans.
3. do you see ban but no mail? what happens instead? do you see some errors in fail2ban.log immediately after restart or more important after ban? If yes, what exactly.
4. `fail2ban-client -d | grep "$jail"` would show you the current dump of config for jail `$jail` (only valid after f2b restart), is there an action `smtp.py` and are all parameters still OK.

> I wanted to avoid installing sendmail or other software.

It looks like a precedence issue between "installing sendmail" and "trying to understand and configure f2b action properly"...
Is the time you (and me) would invest here fewer important than you wish not to install something?
Let alone the usage of mailing actions (especially over foreign smtp) are not recommended at all - it simply slowing down a banning process, may unnecessary retard the bans of others IPs, and can generate extra load on the machine (see #832), etc.
A periodic mail (e. g. by cron) with status of fail2ban/jail is fully enough.
